libraries/AP_OpticalFlow/AP_OpticalFlow_Calibrator.cpp

```diff
@@ -0,0 +1,320 @@
+        _cal_data[0].samples[_cal_data[0].num_samples].flow_rate = flow_rate.x;
+        _cal_data[0].samples[_cal_data[0].num_samples].body_rate = body_rate.x;
+        _cal_data[0].samples[_cal_data[0].num_samples].los_pred = los_pred.x;
+        _cal_data[0].num_samples++;
+    }
+    const bool rates_y_sufficient = (fabsf(body_rate.y) >= AP_OPTICALFLOW_CAL_ROLLPITCH_MIN_RADS) && (fabsf(flow_rate.y) >= AP_OPTICALFLOW_CAL_ROLLPITCH_MIN_RADS);
+    if (rates_y_sufficient && (_cal_data[1].num_samples < ARRAY_SIZE(_cal_data[1].samples))) {
+        log_sample(1, _cal_data[1].num_samples, flow_rate.y, body_rate.y, los_pred.y);
+        _cal_data[1].samples[_cal_data[1].num_samples].flow_rate = flow_rate.y;
+        _cal_data[1].samples[_cal_data[1].num_samples].body_rate = body_rate.y;
+        _cal_data[1].samples[_cal_data[1].num_samples].los_pred = los_pred.y;
+        _cal_data[1].num_samples++;
+    }
+}
+
+// returns true once the sample buffer is full
+bool AP_OpticalFlow_Calibrator::sample_buffers_full() const
+{
+    return ((_cal_data[0].num_samples >= ARRAY_SIZE(_cal_data[0].samples)) && (_cal_data[1].num_samples >= ARRAY_SIZE(_cal_data[1].samples)));
+}
+
+// run calibration algorithm for both axis
+// returns true on success and updates _cal_data[0,1].best_scale and best_scale_fitness
+bool AP_OpticalFlow_Calibrator::run_calibration()
+{
+    // run calibration for x and y axis
+    const bool calx_res = calc_scalars(0, _cal_data[0].best_scalar, _cal_data[0].best_scalar_fitness);
+    const bool caly_res = calc_scalars(1, _cal_data[1].best_scalar, _cal_data[1].best_scalar_fitness);
+
+    return calx_res && caly_res;
+}
+
+// Run Gauss Newton fitting algorithm for all samples of the given axis
+// returns a scalar and fitness (lower numbers mean a better result) in the arguments provided
+bool AP_OpticalFlow_Calibrator::calc_scalars(uint8_t axis, float& scalar, float& fitness)
+{
+    // prefix for reporting
+    const char* prefix_str = "FlowCal:";
+    const char* axis_str = axis == 0 ? "x" : "y";
+
+    // check we have samples
+    // this should never fail because this method should only be called once the sample buffer is full
+    const uint8_t num_samples = _cal_data[axis].num_samples;
+    if (num_samples == 0) {
+        gcs().send_text(MAV_SEVERITY_INFO, "%s failed because no samples", prefix_str);
+        return false;
+    }
+
+    // calculate total absolute residual from all samples
+    float total_abs_residual = 0;
+    for (uint8_t i = 0; i < num_samples; i++) {
+        const sample_t& samplei = _cal_data[axis].samples[i];
+        total_abs_residual += fabsf(calc_sample_residual(samplei, 1.0));
+    }
+
+    // if there are no residuals then scaling is perfect
+    if (is_zero(total_abs_residual)) {
+        scalar = 1.0;
+        fitness = 0;
+        gcs().send_text(MAV_SEVERITY_INFO, "%s perfect scalar%s of 1.0", prefix_str, axis_str);
```

Comment:
need to use GCS_SEND_TEXT() so it is AP_Periph friendly

Reply:
Txs again for the review.  Changed to use GCS_SEND_TEXT.